cudamapper/src/main.cu

```diff
@@ -306,27 +301,54 @@ void process_one_batch(const BatchOfIndices& batch,
 
     // load indices into host memory
     {
+        GW_NVTX_RANGE(profiler, "main::process_one_batch::generte_host_indices");
         assert(!host_batch.query_indices.empty() && !host_batch.target_indices.empty() && !device_batches.empty());
 
-        GW_NVTX_RANGE(profiler, "main::process_one_batch::host_indices");
-        host_cache.generate_query_cache_content(host_batch.query_indices,
+        index_cache.generate_content_query_host(host_batch.query_indices,
                                                 device_batches.front().query_indices,
                                                 skip_copy_to_host);
-        host_cache.generate_target_cache_content(host_batch.target_indices,
+        index_cache.generate_content_target_host(host_batch.target_indices,
                                                  device_batches.front().target_indices,
                                                  skip_copy_to_host);
     }
 
     // process device batches one by one
-    for (const IndexBatch& device_batch : batch.device_batches)
+    // Processing one device batch is overlapped with fetching indices from host to device for the next device batch.
+    // Loop uses copy_device_batch_index as its index, compute_device_batch_index is equal to copy_device_batch_index - 1
+    // (compute_device_batch_index does not actually exist).
+    // copy_device_batch_index loops over one more element because for such element compute_device_batch_index would have been equal to the last batch
+    for (int32_t copy_device_batch_index = 0; copy_device_batch_index < get_size<int32_t>(batch.device_batches) + 1; ++copy_device_batch_index)
```

Comment:
I agree that the object carries a lot of state and it is rather complicated to track. I think better naming could make it clearer, although it would be still far from perfect.

The issue with the solution you suggested is probably the main reason I went for unifying host and device caches in the first place. For sufficiently large datasets as the application moves over batches of indices query indices are going to remain the same, while the target ones change. This means that in many cases we can avoid moving the data to device as it is already there, but host cache has to be aware of that and keep track of currently active device caches. This could be done by having shared pointers to device indices in both host and device caches, but the problem is that host cache has to know when some device indices are not needed anymore and for that it would have to in a way be the owner of device caches or device caches in their destructors would have to let host cache cache that they do not need those indices anymore.

I suggest having a call about this. It is a rather complicated topic and leans towards trading off simplicity for performance, but that still does not mean that we should not have another look at it.